### ISSUE TITLE
doctor's/nurses after resting abandon their consults and wont take new consults.

### ISSUE THREAD

**Issue by overlordchaos** [CorsixTH/CorsixTH#1290]
#### Describe the issue
every time a doctor (surgeon/doctor/psychiatrist/investigation/specialist/nurse) after going to rest, 90% of the times wont go back to their assigned station will just roam around. and very rarely when needed will go to a compatible station (for example if the nurse goes out the enfirmary to rest, then she does not go back, but might run into the pharmacy, those are rare cases like 10% or less) this also happened on the original game. it does not make it less of an issue that should be resolved.
#### Steps to Reproduce
1. run any hospital with several doctors/nurses in

2.let them to work then go to rest. once that is done they will start to roam around instead of go back to their assignated stations.

3. (just obserb)

#### Expected Behavior
they should go back to their stations, and if they are occupied by a new staff member, in the second they are needed in another place, auto assign that station to themselves.
#### System Information
**CorsixTH Version:** *e.g. 0.50 or 7d886f35ca*
6.1 (beta)
**Operating System:** *e.g. Windows 10*
windows 10 x64
**Theme Hospital Version:** *CD, GOG.com, Origin, or Demo*
gog.com
### Additional Info
it could be useful to check the psychiatrist code, considering the psychiatrist (even on the original game) in the second a regular doctor is needed, they usually run to take the place of the regular doctor. this code could be modded and applied to all the staff so this issue does not accur anymore.

**Comment by TheCycoONE** [comment]
Classifying as a low priority enhancement, since micromanaging doctors after the leave for rest was part of the original game; but you're right it's kinda tedious and doesn't add much fun. The alternative was the stick to your room policy, in which case you'd have to micromanage how tired they were getting.

**Comment by overlordchaos** [comment]
yes aldo if you force them to stick to your room, you will have to manually replace the doctor/nurse and when they are rested switch them back. it is quite tedious if you have 30+ staff members and you need to be tracking down each one. aldo i am sure its an original game bug to have with or without the room policy to doctors/nurses roaming around while the recepcionist asks for a replacement in a room and they ignore it while patients wait untill death or leaving angry. its a mix of game mechanic/glitch if you ask me ^_^

**Comment by MarkL1961** [comment]
That is somewhat realistic, there are people where I work who just wander round all day whilst others are run off their feet :)

**Comment by lewri** [comment]
Actually I think might be normal behaviour of the game.

When 'stick to your room' policy applies, doctors and nurses in the first instance should go back to their original rooms. Only if someone else took over the room when they rested would they roam.
When this policy is off, doctors and nurses will wait for a page/call to go to a specific room. The idea of this is to allow everyone to go where they are needed rather than having something permanently stationed.

The idea of this enhancement, if I understand, is that 'stick to your room' policy should always apply but with loose and strict elements. Such that:

**Strict condition:** _Behaviour like old game's mechanic, staff always stick to their rooms unless occupied by another staff member in their absence._

**Loose condition:** _Staff prioritize the same room they were in after resting, but can be called to another room while heading to their room (or meander if not needed)_

While I do not think this is a bad idea I believe the dichotomy of this policy's effect is lost. We could create a 3-step slide to include: _Off, Loose, Strict_ if it is really favoured, however.

**Comment by MarkL1961** [comment]
@lewri  with those rooms if you have more than one they go to either after a break.  This can be annoying if you have one training room to train surgeons and another for psychiatrists as they end up learning both over a longer period.
Having an option where staff go back to their room first would be a bonus

**Comment by joh024** [comment]
While not a direct resolution, I also found similar situations (only 1 room being actively used with queues of 1-2, when there are empty rooms right next door and staff that could fill these rooms doing nothing in the corridor) to be annoying to micro-manage, so I added some extra code to `room.lua` that made things flow much more smoothly.

To preface, the thinking behind the change is this: If the player has built rooms and hasn't disabled them (set queue size to 0), then they expect these rooms to be used.

The new code modifies `Room:getUsageScore` in [room.lua](https://github.com/CorsixTH/CorsixTH/blob/master/CorsixTH/Lua/room.lua) which is a "Score function to decide how desirable a room is for a patient." and uses the following special values.

```lua
local tile_factor = 10     -- how many tiles further are we willing to walk for 1 person fewer in the queue
local readiness_bonus = 50 -- how many tiles further are we willing to walk if the room has all the required staff
```

I add to this an "empty" bonus:

```lua
local empty_bonus = 30     -- how many tiles further are we willing to walk if the room has empty queue
```

Value of 30 is arbitrary, but seemed to work fine for my playthrough.

Then I add the following into `Room:getUsageScore`

```lua
  if queue:patientSize() == 0 then
    score = score - empty_bonus
  end
```

With this change patients are more readily assigned to empty rooms, which leads to a dispatch call to staff, which gets them out of the corridor (well, usually. Sometimes staff from another room that's not currently busy gets called over instead, but that's working as intended) and accomplishes the goal of having rooms/staff being more actively used.

**Comment by joh024** [comment]
Probably worth posting some screenshots to demonstrate it in action.

N.B. For this test I modified `World:updateSpawnDates` in `world.lua` to boost spawn rates, just so we don't have to wait as long.

1. Start a Single Scenario on St. Peter's near Cawley.
2. Immediately pause, build 1 Reception and 3 identical GP's Offices. Hire the required staff, but place them in corridor.
3. Save game.

![base](https://user-images.githubusercontent.com/3042623/34940876-2d376b1e-fa56-11e7-9e63-36fd37915c59.png)

### Without changes

Load game, open hospital then unpause.

![default_0213](https://user-images.githubusercontent.com/3042623/34940883-3698e034-fa56-11e7-9a07-e78f9bad6b71.png)

Already by February 13 we see the problem occurring. Only 1 office is in use with queue of 4, with 2 offices and 2 doctors idle.

![default_0310](https://user-images.githubusercontent.com/3042623/34940886-3776629c-fa56-11e7-8e8f-39d0c6db69cd.png)

By March 10 we've finally got the second office active, but both offices have big queues with third office and doctor still idle.


### With changes

Load game, open hospital then unpause.

![mod_0213](https://user-images.githubusercontent.com/3042623/34940976-9c2fbd82-fa56-11e7-9201-1c333f6649c7.png)

By the same February 13 we see the second office is active, and it has been for a few days. Longest queue at the moment is only size 2.

![mod_0219](https://user-images.githubusercontent.com/3042623/34940977-9cdca65a-fa56-11e7-9cff-16c3edf0f893.png)

By February 19 we have all 3 offices active. I redid this a couple of times and the third office was active by around Feb 16-20, depending on patient inflow and where the doctor had wandered off to (since I only counted it active when the doctor was inside and working).

Pretty simple change but with noticeable results, and a very nice quality of life change that lowers annoying micro.

**Comment by MarkL1961** [comment]
If you were to reduce the max queue size for each room the other rooms would get used sooner too

**Comment by TheCycoONE** [comment]
@joh024 I think your code doesn't match the issue. I also think it's worth considering. Could you look if there is a more suitable issue, and if not then create a new one?

**Comment by LeDocteurr** [comment]
I fixed this issue in PR #1405 , when "staff leave rooms" is disabled the staff will stick to their rooms. Further enhancements could be done so specialists may return to their preferred posts even when the option is enabled, but that will reduce the micromanagement, so it's debatable.

**Comment by lewri** [comment]
There was a fix for this but wasn't used. Should be be cleaned up to favour #1582 ?

**Comment by mugmuggy** [comment]
#1582 is relating to a doctor already in the office and could be considered as related to this issue - its actually similar behaviour to what is desired but is actually a separate issue.

The MeanderAction after exiting the staff room already checks to see if the staff member should return to the previous post (it has to be free). Otherwise they meander until a job comes up.

This is the gripe from the original post. TH just allocates once and does away with it, but Corsix allows calls to be reallocated to a doctor that becomes available also and is considered closer. It also prefers tired doctors over freshly recovered ones. But the looking for jobs doesn't trump this and thus you potentially get Drs leaving a room walking past a wandering Dr to answer a call that they are just as close to, hence #1600 occurs as they don't get called to answer something.

For example Staff.last_room will be nil for staff that have exited the staff room and can't go back to previous room but you might turn too far away from the micromanagement side if you blanketly apply it. Distance especially across custom maps where the hospital layout might dictate long paths, means that distance isn't a reliable offsetting factor. Might be able to get them to return to another room of that type if there is one free for example, so left a GP, find a GP. There is still a chance a call will go out and another Dr anwers it though.

The other one that I noticed is for epidemics and thus nurses in TH tend to chase more and will pull up on a patient rather than wait for them to be completely idle. And I think that is related to how TH calculates paths also with particular points of recalculation for the path to occur are opportunities.
